### PR TITLE
[dynamo] Add graph break on requires_grad_()

### LINKED ISSUE #107861: torch._dynamo.optimize loses the setting of requires_grad
### 🐛 Describe the bug

The issue is exposed by https://github.com/pytorch/pytorch/actions/runs/5917745793/job/16046614052.
When torch._dynamo.optimize is enabled for the code shown below,  the setting of `requires_grad` is lost.

```
import torch

def traverse_obj(obj, func):
    if isinstance(obj, (tuple, list)):
        a = type(obj)(traverse_obj(o, func) for o in obj)
        print("return tuple tensor: ", a[0].requires_grad)
        return a
    elif isinstance(obj, dict):
        return {name: traverse_obj(o, func) for name, o in obj.items()}
    elif isinstance(obj, (torch.Tensor, torch.nn.Parameter)):
        b = func(obj)
        print("return tensor: ", b.requires_grad)
        return b

def to_mem_format(mem_format, inputs):
    def inner_to_mem_format(obj):
        old_requires_grad = obj.requires_grad
        _tensor = obj.clone().to(memory_format=mem_format).detach().requires_grad_(old_requires_grad)
        print("inner cloned tensor : ", _tensor.requires_grad)
        return _tensor
    print("input requires_grad: ", inputs[0].requires_grad)
    return traverse_obj(inputs, inner_to_mem_format)


super_run = torch._dynamo.optimize("aot_eager")(to_mem_format)

inputs = (torch.randn(4, 3, 4, 4).requires_grad_(True),)

output = super_run(torch.channels_last, inputs)

print("output requires_grad: ", output[0].requires_grad)
```

Resutls:

```
input requires_grad:  True
inner cloned tensor :  True
return tensor:  False
return tuple tensor:  False
output requires_grad:  False
```

### Versions

Latest master.

cc @ezyang @gchanan @zou3519 @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### PR DESCRIPTION
**Pull request by int3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110053

Fixes #107861.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -690,6 +689,8 @@ def redistribute_fn_with_prim_types(x):
                 ), "Unreachable - See unimplemented for lambdas above"
             tx.output.side_effects.register_hook(self, fn_var, handle_variable)
             return handle_variable
+        elif name == "requires_grad_":
+            unimplemented("Tensor.requires_grad_")
```
Comment (eellison) [inline torch/_dynamo/variables/tensor.py:693]: Should we avoid graph-breaking on the case when this is a no-op? If tensor already requires grad and the input is True, or vice versa

**Review by eellison — COMMENTED** [review]
cool! one comment. cc @bdhirsh, we graph-broke for resize_ only when the tensor was a graph input, since it's complicated to replay the logic. i imagine for requires_grad we will want to graph break on both inputs and intermediaries because aot_autograd does not handle intermediaries with gradients currently. want to confirm?

File: torch/_dynamo/variables/tensor.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_dynamo/variables/tensor.py:693]: +1

File: test/test_ops.py [hunk 2]
```diff
@@ -1622,7 +1622,6 @@ def clone_and_perform_view(input, **kwargs):
                         self.assertEqual(tensor.grad, cloned1_tensor.grad)
 
     @ops(ops_and_refs, allowed_dtypes=(torch.cfloat,))
-    @skipIfTorchInductor("Inductor does not support complex dtype yet")
```
Comment (bdhirsh) [inline test/test_ops.py:1625]: do we know why this impacted complex dtype tests?

Comment (int3) [reply test/test_ops.py:1625]: I'm not sure what the complex dtype issue was originally but I think it was already fixed. `skipIfTorchInductor` is a bit of a heavy hammer, it skips all variants of a test even if only a subset is failing under inductor...

I thought the fact that the `test_neg_view_cuda` tests are now passing was kinda weird though. I'm surprised it's *just* the CUDA tests, and that the corresponding CUDA tests for `test_conj_view` and `test_neg_cong_view` weren't similarly affected. Do you want me to find the root cause? Might take a while

**Comment by int3** [comment]
bump

**Comment by eellison** [comment]
@int3 next time if you wouldnt mind click this here when you're ready for another review 
image

File: test/test_ops.py (hunk 2 above)
Comment (eellison) [reply test/test_ops.py:1625]: It wasn't anything to do with complex - it was just this logic [here](https://github.com/pytorch/pytorch/blob/437b1aae4a693540a9da9c1cbca8f9b22a38a8cd/test/test_ops.py#L1550) to create the tensor that affected the tests

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -9636,7 +9642,9 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
            skips=(
                # Tests that assume input tensor has a meaningful effect on output tensor
                DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_variant_consistency_eager'),
-               DecorateInfo(unittest.expectedFailure, 'TestMathBits', 'test_neg_view'),
+               DecorateInfo(unittest.expectedFailure, 'TestMathBits', 'test_neg_view', device_type='cpu'),
+               DecorateInfo(unittest.expectedFailure, 'TestMathBits', 'test_neg_view', device_type='cuda',
```
Comment (eellison) [inline torch/testing/_internal/common_methods_invocations.py:9646]: So the test should fail on cuda unless we're running with TEST_WITH_TORCHINDUCTOR 🤔 that seems a bit odd

**Review by eellison — APPROVED** [review]
looks good, thanks!

**Comment by bdhirsh** [comment]
> we graph-broke for resize_ only when the tensor was a graph input, since it's complicated to replay the logic. i imagine for requires_grad we will want to graph break on both inputs and intermediaries because aot_autograd does not handle intermediaries with gradients currently. want to confirm?

Sorry, missed this comment. yep - we'll want to graph break on both. In theory, the input case is probably easier to handle, and we could support it in AOTAutograd. Seems safer to graph break on both for now though, and try to fix it later if someone complains (since it should be pretty easy to re-write your model code to avoid the graph break)

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (int3) [reply torch/testing/_internal/common_methods_invocations.py]: Whatever the cause was, it seems to have gone away after rebasing (i.e. the test now uniformly fails regardless of `TEST_WITH_TORCHINDUCTOR`...)
